Analisis Exploratorio de Datos de los datasets Coursera, Edx y Udemy


Importamos librerias necesarias para el Analisis 

In [1]:
import pandas as pd #libreria para trabajar con dataframes
from langdetect import detect #libreria para detectar idiomas

Según el objetivo propuesto en el trabajo se busca segmentar la información a nivel de ventas según precio, idioma, nivel del curso y el rating. Por lo que se procede a remover aquellas columnas de los datasets que no formen parte de la consigna.

In [238]:
# Cargamos dataset Coursera_courses.csv
data_coursera_courses = pd.read_csv('./Coursera_courses.csv')

                                            COURSERA COURSES EDA

In [239]:
data_coursera_courses.head()

name  \
0                                   Machine Learning   
1                                  Indigenous Canada   
2                          The Science of Well-Being   
3                     Technical Support Fundamentals   
4  Become a CBRS Certified Professional Installer...   

                 institution  \
0        Stanford University   
1      University of Alberta   
2            Yale University   
3                     Google   
4  Google - Spectrum Sharing   

                                          course_url  \
0    https://www.coursera.org/learn/machine-learning   
1   https://www.coursera.org/learn/indigenous-canada   
2  https://www.coursera.org/learn/the-science-of-...   
3  https://www.coursera.org/learn/technical-suppo...   
4  https://www.coursera.org/learn/google-cbrs-cpi...   

                        course_id  
0                machine-learning  
1               indigenous-canada  
2       the-science-of-well-being  
3  technical-support-fundamentals  
4        google-cbrs-cpi-training

Análisis exploratorio para conocer la estructura de los datos y ver si hay errores.

In [27]:
data_coursera_courses.info()
data_coursera_courses.value_counts()
data_coursera_courses.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 623 entries, 0 to 622
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         623 non-null    object
 1   institution  623 non-null    object
 2   course_url   623 non-null    object
 3   course_id    623 non-null    object
dtypes: object(4)
memory usage: 19.6+ KB


name           0
institution    0
course_url     0
course_id      0
dtype: int64

Chequeamos por valores duplicados.

In [28]:
data_coursera_courses.duplicated().value_counts()

False    623
dtype: int64

Me quedo con las columnas que tengan relevancia para el análisis.

In [70]:
data_coursera_courses_corregido = data_coursera_courses[['name','course_id']]

In [72]:
data_coursera_courses_corregido

name  \
0                                     Machine Learning   
1                                    Indigenous Canada   
2                            The Science of Well-Being   
3                       Technical Support Fundamentals   
4    Become a CBRS Certified Professional Installer...   
..                                                 ...   
618              Accounting Data Analytics with Python   
619             Introduction to Molecular Spectroscopy   
620                                Managing as a Coach   
621             The fundamentals of hotel distribution   
622                     A Crash Course in Data Science   

                            course_id  
0                    machine-learning  
1                   indigenous-canada  
2           the-science-of-well-being  
3      technical-support-fundamentals  
4            google-cbrs-cpi-training  
..                                ...  
618  accounting-data-analytics-python  
619                      spectroscopy  
620               managing-as-a-coach  
621                hotel-distribution  
622               data-science-course  

[623 rows x 2 columns]

NO SE ENCONTRARON DATOS NULOS NI DUPLICADOS EN EL DATASET COURSERA_COURSES.CSV

-------------------------------------------------------------------------------------------

                                            COURSERA REVIEWS EDA

In [80]:
# Cargamos dataset Coursera_reviews.csv
data_coursera_reviews = pd.read_csv('./Coursera_reviews.csv')

In [82]:
data_coursera_reviews.head()

reviews       reviewers  \
0  Pretty dry, but I was able to pass with just t...     By Robert S   
1  would be a better experience if the video and ...  By Gabriel E R   
2  Information was perfect! The program itself wa...      By Jacob D   
3  A few grammatical mistakes on test made me do ...       By Dale B   
4  Excellent course and the training provided was...       By Sean G   

   date_reviews  rating                 course_id  
0  Feb 12, 2020       4  google-cbrs-cpi-training  
1  Sep 28, 2020       4  google-cbrs-cpi-training  
2  Apr 08, 2020       4  google-cbrs-cpi-training  
3  Feb 24, 2020       4  google-cbrs-cpi-training  
4  Jun 18, 2020       4  google-cbrs-cpi-training

Análisis exploratorio para conocer la estructura de los datos y ver si hay errores


In [83]:
data_coursera_reviews.info()
data_coursera_reviews.value_counts()
data_coursera_reviews.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1454711 entries, 0 to 1454710
Data columns (total 5 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   reviews       1454571 non-null  object
 1   reviewers     1454711 non-null  object
 2   date_reviews  1454711 non-null  object
 3   rating        1454711 non-null  int64 
 4   course_id     1454711 non-null  object
dtypes: int64(1), object(4)
memory usage: 55.5+ MB


reviews         140
reviewers         0
date_reviews      0
rating            0
course_id         0
dtype: int64

Chequeo por duplicados

In [86]:
data_coursera_reviews.duplicated().value_counts()

True     934764
False    519947
dtype: int64

In [87]:
data_coursera_reviews.drop_duplicates(inplace=True) #Remuevo duplicados

In [88]:
data_coursera_reviews.dropna(inplace=True) #Remuevo nulos

Me quedo con las columnas que tengan relevancia para el análisis y coloco todo el un nuevo dataframe donde realizare otras transformaciones.

In [179]:
data_coursera_reviews_corregido = data_coursera_reviews[['rating','course_id']]

Reviso cantidad de registros en base a su course_id

In [180]:
len(data_coursera_reviews_corregido['course_id'].unique()) 

604

Diferentes tipos de ratings

In [181]:
data_coursera_reviews_corregido['rating'].unique()

array([4, 5, 1, 2, 3])

Creo columna que permite ver Número de reviews. Para ellos hago una agrupación en base al 'course_id' y realizo una transformación para contar todos los registros por cada curso por separado.

In [183]:
data_coursera_reviews_corregido['n_reviews'] = data_coursera_reviews_corregido.groupby('course_id')['course_id'].transform('count')

/var/folders/4n/z_dkq44j2t9bgj_4nq5l3m4h0000gn/T/ipykernel_57233/3839990971.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_coursera_reviews_corregido['n_reviews'] = data_coursera_reviews_corregido.groupby('course_id')['course_id'].transform('count')


Visualizo el nuevo dataframe con la columna 'n_reviews'

In [208]:
data_coursera_reviews_corregido

rating                 course_id  n_reviews
0             4  google-cbrs-cpi-training         33
1             4  google-cbrs-cpi-training         33
2             4  google-cbrs-cpi-training         33
3             4  google-cbrs-cpi-training         33
4             4  google-cbrs-cpi-training         33
...         ...                       ...        ...
1454636       5       computer-networking       4811
1454639       5       computer-networking       4811
1454642       5       computer-networking       4811
1454643       5       computer-networking       4811
1454644       5       computer-networking       4811

[519891 rows x 3 columns]

Voy sumar la cantidad de votos de rating por curso y luego dividirla por su 'n_reviews' Numero de reviews, para sacar un promedio estimativo del rating del curso.

In [215]:
data_coursera_reviews_corregido['promedio_rating'] = data_coursera_reviews_corregido.groupby('course_id')['rating'].transform('mean')

/var/folders/4n/z_dkq44j2t9bgj_4nq5l3m4h0000gn/T/ipykernel_57233/732755657.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_coursera_reviews_corregido['promedio_rating'] = data_coursera_reviews_corregido.groupby('course_id')['rating'].transform('mean')


In [218]:
data_coursera_reviews_corregido

rating                 course_id  n_reviews  promedio_rating
0             4  google-cbrs-cpi-training         33         4.818182
1             4  google-cbrs-cpi-training         33         4.818182
2             4  google-cbrs-cpi-training         33         4.818182
3             4  google-cbrs-cpi-training         33         4.818182
4             4  google-cbrs-cpi-training         33         4.818182
...         ...                       ...        ...              ...
1454636       5       computer-networking       4811         4.628767
1454639       5       computer-networking       4811         4.628767
1454642       5       computer-networking       4811         4.628767
1454643       5       computer-networking       4811         4.628767
1454644       5       computer-networking       4811         4.628767

[519891 rows x 4 columns]

Continuo a eliminar la columna 'rating' debido a que consegui el promedio por curso

In [219]:
data_coursera_reviews_corregido = data_coursera_reviews_corregido.drop(columns=['rating'])

Continuo a eliminar los duplicados para dejarme únicamente con los registros de los cursos y su promedio de rating. Todo con el fin de disminuir el dataset en una cantidad de registros con los que pueda trabajar mas adelante.

In [222]:
data_coursera_reviews_corregido.drop_duplicates(inplace=True)

Al encontrar 604 registros de course_id en el dataset 'Coursera_reviews.csv' y sabiendo que hay 623 registros de course_id en el dataset 'Coursera_courses.csv'; llegamos a la conclusión de que 19 'course_id' no contienen un 'rating'.

Debido a que contamos con dos datasets para la misma plataforma de Coursera. Continuo a unirlos para poder usar la misma información para los mismos cursos.

In [224]:
data_coursera = pd.merge(data_coursera_courses_corregido,data_coursera_reviews_corregido, how='outer')

Reviso el nuevo dataset que contiene el Titulo del curso, su ID, el Número de reviews y el promedio de rating que obtuvo. Varios registros no tuvieron reviews por lo que sus valores Nan fueron reemplazados por 0.

In [5]:
data_coursera.fillna(0)

Unnamed: 0                                               name  \
0             0                                   Machine Learning   
1             1                                  Indigenous Canada   
2             2                          The Science of Well-Being   
3             3                     Technical Support Fundamentals   
4             4  Become a CBRS Certified Professional Installer...   
..          ...                                                ...   
618         618              Accounting Data Analytics with Python   
619         619             Introduction to Molecular Spectroscopy   
620         620                                Managing as a Coach   
621         621             The fundamentals of hotel distribution   
622         622                     A Crash Course in Data Science   

                            course_id  n_reviews  promedio_rating  
0                    machine-learning    12677.0         4.739923  
1                   indigenous-canada     1460.0         4.816438  
2           the-science-of-well-being     8199.0         4.929504  
3      technical-support-fundamentals    12054.0         4.767546  
4            google-cbrs-cpi-training       33.0         4.818182  
..                                ...        ...              ...  
618  accounting-data-analytics-python        8.0         3.500000  
619                      spectroscopy      408.0         4.607843  
620               managing-as-a-coach      282.0         4.734043  
621                hotel-distribution        0.0         0.000000  
622               data-science-course        0.0         0.000000  

[623 rows x 5 columns]

Luego de revisar el sitio web de la plataforma Coursera('www.coursera.org'); es necesaria una subscripción mensual o anual para poder tomar las clases de los cursos que ofrecen certificados. También se encuentran diferentes tipos de cursos gratis los cuales pueden ser accedidos sin una necesidad de una membresía.


In [3]:
data_coursera = pd.read_csv('data_coursera_revisado.csv')

Con la ayuda de la librería de Langdetect puedo detectar el idioma en el cual estan escritos los titulos de los cursos de Coursera.

https://pypi.org/project/langdetect/
para mas información en la libreria.


Creo una función que ayuda a detectar el lenguaje de cada curso respecto al titulo.

In [7]:
def detectlang(text):
    try:
        return detect(text)
    except:
        return 'Unknown'


In [8]:
data_coursera['idioma'] = data_coursera['name'].apply(detectlang)

In [11]:
data_coursera

Unnamed: 0                                               name  \
0             0                                   Machine Learning   
1             1                                  Indigenous Canada   
2             2                          The Science of Well-Being   
3             3                     Technical Support Fundamentals   
4             4  Become a CBRS Certified Professional Installer...   
..          ...                                                ...   
618         618              Accounting Data Analytics with Python   
619         619             Introduction to Molecular Spectroscopy   
620         620                                Managing as a Coach   
621         621             The fundamentals of hotel distribution   
622         622                     A Crash Course in Data Science   

                            course_id  n_reviews  promedio_rating idioma  
0                    machine-learning    12677.0         4.739923     en  
1                   indigenous-canada     1460.0         4.816438     id  
2           the-science-of-well-being     8199.0         4.929504     en  
3      technical-support-fundamentals    12054.0         4.767546     de  
4            google-cbrs-cpi-training       33.0         4.818182     en  
..                                ...        ...              ...    ...  
618  accounting-data-analytics-python        8.0         3.500000     en  
619                      spectroscopy      408.0         4.607843     en  
620               managing-as-a-coach      282.0         4.734043     tl  
621                hotel-distribution        NaN              NaN     en  
622               data-science-course        NaN              NaN     en  

[623 rows x 6 columns]

Continuo a realizar una tranformación de la columna idioma para que sea mas facil de interpretar.
Mas información en: https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes para conocer los resultados de la libreria Langdetect



In [14]:
data_coursera['idioma'].value_counts()

en    521
es     26
fr     13
de     12
ca     11
it     11
no      7
ru      5
tl      4
nl      3
id      2
da      2
pt      2
ro      2
lt      1
af      1
Name: idioma, dtype: int64

Realizo una tranformacion de los promedio conseguidos para tener un numero aproximado del mismo.

In [17]:
data_coursera['promedio_rating'].sort_values(ascending=False)

106    5.000
500    5.000
603    5.000
50     5.000
361    4.968
       ...  
529      NaN
557      NaN
569      NaN
621      NaN
622      NaN
Name: promedio_rating, Length: 623, dtype: float64

Antes de exportar el dataset decidí ordenarlos en base al numero de reviews con respecto a su promedio de rating.
Continuo a exportar el nuevo dataset.

In [12]:
data_coursera.to_csv('data_coursera_revisado.csv')

In [6]:
data_coursera.drop(columns='Unnamed: 0', inplace=True)

In [14]:
data_coursera.sort_values(['n_reviews','promedio_rating'],ascending=False,inplace=True)

In [230]:
data_coursera

name  \
7    Programming for Everybody (Getting Started wit...   
0                                     Machine Learning   
12                   Neural Networks and Deep Learning   
3                       Technical Support Fundamentals   
24   Learning How to Learn: Powerful mental tools t...   
..                                                 ...   
529            Digital Marketing Analytics in Practice   
557                    Dairy Production and Management   
569  Security & Safety Challenges in a Globalized W...   
621             The fundamentals of hotel distribution   
622                     A Crash Course in Data Science   

                            course_id  n_reviews  promedio_rating  idioma  
7                              python    15226.0         4.583673  Ingles  
0                    machine-learning    12677.0         4.739923  Ingles  
12      neural-networks-deep-learning    12291.0         4.771540  Ingles  
3      technical-support-fundamentals    12054.0         4.767546  Ingles  
24              learning-how-to-learn    11871.0         4.801196  Ingles  
..                                ...        ...              ...     ...  
529                 digital-analytics        0.0         0.000000  Ingles  
557                  dairy-production        0.0         0.000000  Ingles  
569  security-safety-globalized-world        0.0         0.000000  Ingles  
621                hotel-distribution        0.0         0.000000  Ingles  
622               data-science-course        0.0         0.000000  Ingles  

[623 rows x 5 columns]

Procedo a cargar un Dataset el cual contiene información respectiva al Rating y a la Dificultad de los cursos de la plataforma de coursera

In [ ]:
data_coursera_kaggle = pd.read_csv('coursera-course-detail-data.csv')

In [266]:
data_coursera_kaggle.head()

Unnamed: 0                                               name  \
0           0                        "Making" Progress Teach-Out   
1           1         (Business Writing) الكتابة في مجال الأعمال   
2           2  (Giving Helpful Feedback) إعطاء الملاحظات المفيدة   
3           3    (Successful Presentation) العرض التقديمي الناجح   
4           4                                 .Net平台下的软件开发技术毕业项目   

                                                 Url Rating      Difficulty  \
0          https://coursera.org/learn/makingprogress   None  Beginner Level   
1  https://coursera.org/learn/writing-for-busines...   None  Beginner Level   
2             https://coursera.org/learn/feedback-ar    4.8            None   
3  https://coursera.org/learn/presentation-skills-ar    4.9  Beginner Level   
4  https://coursera.org/learn/net-ruanjian-kaifa-...   None            None   

                                            Tags  
0  ['Social Sciences', 'Governance and Society']  
1            ['Business', 'Business Essentials']  
2            ['Business', 'Business Essentials']  
3            ['Business', 'Business Essentials']  
4   ['Computer Science', 'Software Development']

Unifico esta información con el ultimo dataframe que cree con la información de Coursera.

In [ ]:
data_coursera_final = pd.merge(data_coursera,data_coursera_kaggle, how='outer')

In [267]:
data_coursera_final.head()

name  \
0  Programming for Everybody (Getting Started wit...   
1                                   Machine Learning   
2                  Neural Networks and Deep Learning   
3                     Technical Support Fundamentals   
4  Learning How to Learn: Powerful mental tools t...   

                        course_id  n_reviews  idioma Rating  \
0                          python    15226.0  Ingles    4.8   
1                machine-learning    12677.0  Ingles    4.9   
2   neural-networks-deep-learning    12291.0  Ingles    4.9   
3  technical-support-fundamentals    12054.0  Ingles    4.8   
4           learning-how-to-learn    11871.0  Ingles    4.8   

           Difficulty  
0           All Level  
1           All Level  
2  Intermediate Level  
3      Beginner Level  
4           All Level

Debido a que algunos cursos no contenian información con respecto al número de reviews se ve reflejado en un Rating de 0.

Este ultimo archivo CSV contiene la informacion de los estudiantes inscriptos en las diferentes clases de Coursera

In [324]:
data_coursera_git = pd.read_csv('UCoursera_Courses.csv')

In [325]:
data_coursera_git.head()

Unnamed: 0                                       course_title  \
0         134  (ISC)² Systems Security Certified Practitioner...   
1         743  A Crash Course in Causality:  Inferring Causal...   
2         874                     A Crash Course in Data Science   
3         413                            A Law Student's Toolkit   
4         635                A Life of Happiness and Fulfillment   

          course_organization course_Certificate_type  course_rating  \
0                      (ISC)²          SPECIALIZATION            4.7   
1  University of Pennsylvania                  COURSE            4.7   
2    Johns Hopkins University                  COURSE            4.5   
3             Yale University                  COURSE            4.7   
4   Indian School of Business                  COURSE            4.8   

  course_difficulty course_students_enrolled  
0          Beginner                     5.3k  
1      Intermediate                      17k  
2             Mixed                     130k  
3             Mixed                      91k  
4             Mixed                     320k

Elimino columnas que no me sean necesarias en el analisis

In [327]:
data_coursera_git.drop(columns=['Unnamed: 0','course_organization','course_Certificate_type'],inplace=True)

In [335]:
data_coursera_git_final = pd.merge(data_coursera_git,data_coursera_final)

In [338]:
data_coursera_git_final.drop(columns=['Rating','n_reviews','course_id','dificultad'],inplace=True)

In [ ]:
data_coursera_git_final.isna

Procedo a exportar el dataset final revisado para utilizarlo en el dashboard

In [334]:
data_coursera_final

course_title  \
0    Programming for Everybody (Getting Started wit...   
1                                     Machine Learning   
2                    Neural Networks and Deep Learning   
3                       Technical Support Fundamentals   
4    Learning How to Learn: Powerful mental tools t...   
..                                                 ...   
621            Digital Marketing Analytics in Practice   
622                    Dairy Production and Management   
623  Security & Safety Challenges in a Globalized W...   
624             The fundamentals of hotel distribution   
625                     A Crash Course in Data Science   

                            course_id  n_reviews  idioma Rating  \
0                              python    15226.0  Ingles    4.8   
1                    machine-learning    12677.0  Ingles    4.9   
2       neural-networks-deep-learning    12291.0  Ingles    4.9   
3      technical-support-fundamentals    12054.0  Ingles    4.8   
4               learning-how-to-learn    11871.0  Ingles    4.8   
..                                ...        ...     ...    ...   
621                 digital-analytics        0.0  Ingles      0   
622                  dairy-production        0.0  Ingles    4.8   
623  security-safety-globalized-world        0.0  Ingles    4.7   
624                hotel-distribution        0.0  Ingles    4.7   
625               data-science-course        0.0  Ingles    4.5   

             dificultad Plataforma  
0             All Level   Coursera  
1             All Level   Coursera  
2    Intermediate Level   Coursera  
3        Beginner Level   Coursera  
4             All Level   Coursera  
..                  ...        ...  
621           All Level   Coursera  
622           All Level   Coursera  
623      Beginner Level   Coursera  
624           All Level   Coursera  
625           All Level   Coursera  

[626 rows x 7 columns]

In [268]:
data_coursera_git_final.to_csv('data_coursera_revisado.csv')

-------------------------------------------------------------------------------------------

                                                    EDX EDA

Cargamos dataset edx_courses.csv


In [93]:
data_edx = pd.read_csv('./edx_courses.csv')

In [94]:
data_edx.head()

title  \
0                                How to Learn Online   
1  Programming for Everybody (Getting Started wit...   
2            CS50's Introduction to Computer Science   
3                                 The Analytics Edge   
4  Marketing Analytics: Marketing Measurement Str...   

                                             summary n_enrolled  \
0  Learn essential strategies for successful onli...    124,980   
1  This course is a "no prerequisite" introductio...    293,864   
2  An introduction to the intellectual enterprise...  2,442,271   
3  Through inspiring examples and stories, discov...    129,555   
4     This course is part of a MicroMasters® Program     81,140   

                           course_type                            institution  \
0              Self-paced on your time                                    edX   
1              Self-paced on your time             The University of Michigan   
2              Self-paced on your time                     Harvard University   
3  Instructor-led on a course schedule  Massachusetts Institute of Technology   
4              Self-paced on your time     University of California, Berkeley   

                                         instructors         Level  \
0            Nina Huntemann-Robyn Belair-Ben Piscopo  Introductory   
1                                  Charles Severance  Introductory   
2                 David J. Malan-Doug Lloyd-Brian Yu  Introductory   
3  Dimitris Bertsimas-Allison O'Hair-John Silberh...  Intermediate   
4                                     Stephan Sorger  Introductory   

                        subject language subtitles         course_effort  \
0  Education & Teacher Training  English   English    2–3 hours per week   
1              Computer Science  English   English    2–4 hours per week   
2              Computer Science  English   English   6–18 hours per week   
3    Data Analysis & Statistics  English   English  10–15 hours per week   
4              Computer Science  English   English    5–7 hours per week   

  course_length                                         price  \
0       2 Weeks   FREE-Add a Verified Certificate for $49 USD   
1       7 Weeks   FREE-Add a Verified Certificate for $49 USD   
2      12 Weeks   FREE-Add a Verified Certificate for $90 USD   
3      13 Weeks  FREE-Add a Verified Certificate for $199 USD   
4       4 Weeks  FREE-Add a Verified Certificate for $249 USD   

                                  course_description  \
0  Designed for those who are new to elearning, t...   
1  This course aims to teach everyone the basics ...   
2  This is CS50x , Harvard University's introduct...   
3  In the last decade, the amount of data availab...   
4  Begin your journey in a new career in marketin...   

                                     course_syllabus  \
0  Welcome - We start with opportunities to meet ...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                          course_url  
0     https://www.edx.org/course/how-to-learn-online  
1  https://www.edx.org/course/programming-for-eve...  
2  https://www.edx.org/course/cs50s-introduction-...  
3      https://www.edx.org/course/the-analytics-edge  
4  https://www.edx.org/course/marketing-analytics...

Análisis exploratorio para conocer la estructura de los datos y ver si hay errores.

In [95]:
data_edx.info()
data_edx.value_counts()
data_edx.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 975 entries, 0 to 974
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   title               975 non-null    object
 1   summary             922 non-null    object
 2   n_enrolled          855 non-null    object
 3   course_type         975 non-null    object
 4   institution         975 non-null    object
 5   instructors         972 non-null    object
 6   Level               975 non-null    object
 7   subject             975 non-null    object
 8   language            975 non-null    object
 9   subtitles           972 non-null    object
 10  course_effort       975 non-null    object
 11  course_length       975 non-null    object
 12  price               975 non-null    object
 13  course_description  935 non-null    object
 14  course_syllabus     417 non-null    object
 15  course_url          975 non-null    object
dtypes: object(16)
memory usage

title                   0
summary                53
n_enrolled            120
course_type             0
institution             0
instructors             3
Level                   0
subject                 0
language                0
subtitles               3
course_effort           0
course_length           0
price                   0
course_description     40
course_syllabus       558
course_url              0
dtype: int64

Chequeo duplicados

In [99]:
data_edx.duplicated().value_counts()

False    974
True       1
dtype: int64

SE ENCONTRARON VALORES NULOS Y UN VALOR DUPLICADO.
APLICAMOS UNA LIMPIEZA DE LOS DATOS Y GUARDAMOS EL RESULTADO EN OTRO DATAFRAME

In [100]:
data_edx.drop_duplicates(inplace=True) #Eliminamos valores duplicados

Me quedo con las columnas que tengan relevancia para el análisis y las guardo en un duplicado para asi conservar el original.


In [230]:
data_edx_corregido = data_edx[['title','price','n_enrolled','Level','language','subject']]

In [231]:
data_edx_corregido.head()

title  \
0                                How to Learn Online   
1  Programming for Everybody (Getting Started wit...   
2            CS50's Introduction to Computer Science   
3                                 The Analytics Edge   
4  Marketing Analytics: Marketing Measurement Str...   

                                          price n_enrolled         Level  \
0   FREE-Add a Verified Certificate for $49 USD    124,980  Introductory   
1   FREE-Add a Verified Certificate for $49 USD    293,864  Introductory   
2   FREE-Add a Verified Certificate for $90 USD  2,442,271  Introductory   
3  FREE-Add a Verified Certificate for $199 USD    129,555  Intermediate   
4  FREE-Add a Verified Certificate for $249 USD     81,140  Introductory   

  language                       subject  
0  English  Education & Teacher Training  
1  English              Computer Science  
2  English              Computer Science  
3  English    Data Analysis & Statistics  
4  English              Computer Science

Reviso valores nulos de la columna 'n_enrolled'

In [251]:
data_edx_corregido['n_enrolled'].isna().value_counts()

False    854
True     120
Name: n_enrolled, dtype: int64

Rellenamos los valores de la columna 'n_enrolled' con 0 para que no genere problemas en el análisis.

In [256]:
data_edx_corregido['n_enrolled'].fillna(0)

0        124,980
1        293,864
2      2,442,271
3        129,555
4         81,140
         ...    
970            0
971            0
972       11,246
973        8,775
974            0
Name: n_enrolled, Length: 974, dtype: object

Reviso valores unicos.

In [130]:
data_edx_corregido['Level'].unique()

array(['Introductory', 'Intermediate', 'Advanced'], dtype=object)

In [131]:
data_edx_corregido['language'].unique()

array(['English', 'Español', 'Italiano', '日本語', 'Français', '中文',
       'Português', 'اللغة العربية', 'Deutsch'], dtype=object)

Retiramos la información de la columa 'price' para quedarnos unicamente con datos numericos.
LLegamos a la conclusión de que todos los cursos propuesto en el dataset de EDX son gratis y que también permiten agregar un certificado verificado por un cierto precio valorizado en dolares americanos. '$'

In [ ]:
data_edx_corregido['price'].unique()

In [245]:
data_edx_corregido['price'] = data_edx_corregido['price'].str.replace(r'FREE-Add a Verified Certificate for USD', '')

/var/folders/4n/z_dkq44j2t9bgj_4nq5l3m4h0000gn/T/ipykernel_57233/3328716685.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_edx_corregido['price'] = data_edx_corregido['price'].str.replace(r'FREE-Add a Verified Certificate for USD', '')


In [247]:
data_edx_corregido = pd.read_csv('data_edx_revisado.csv')

In [249]:
data_edx_corregido.drop(columns='Unnamed: 0',inplace=True)

In [251]:
data_edx_corregido['price'] = data_edx_corregido['price'].str.replace(r'$', '')

/var/folders/4n/z_dkq44j2t9bgj_4nq5l3m4h0000gn/T/ipykernel_7871/1658196412.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data_edx_corregido['price'] = data_edx_corregido['price'].str.replace(r'$', '')


In [252]:
data_edx_corregido

title price n_enrolled  \
0                                  How to Learn Online   49     124,980   
1    Programming for Everybody (Getting Started wit...   49     293,864   
2              CS50's Introduction to Computer Science   90   2,442,271   
3                                   The Analytics Edge  199     129,555   
4    Marketing Analytics: Marketing Measurement Str...  249      81,140   
..                                                 ...   ...        ...   
969  Leaders in Citizen Security and Justice Manage...   25         NaN   
970              Pattern Studying and Making | 图案审美与创作  139         NaN   
971  Computational Neuroscience: Neuronal Dynamics ...  139      11,246   
972  Cities and the Challenge of Sustainable Develo...   25       8,775   
973                      MathTrackX: Special Functions   79         NaN   

            Level language                       subject  
0    Introductory  English  Education & Teacher Training  
1    Introductory  English              Computer Science  
2    Introductory  English              Computer Science  
3    Intermediate  English    Data Analysis & Statistics  
4    Introductory  English              Computer Science  
..            ...      ...                           ...  
969  Intermediate  English               Social Sciences  
970  Introductory       中文                 Art & Culture  
971      Advanced  English       Biology & Life Sciences  
972  Introductory  English         Environmental Studies  
973  Introductory  English                          Math  

[974 rows x 6 columns]

In [257]:
data_edx_corregido.to_csv('data_edx_revisado.csv')

-------------------------------------------------------------------------------------------

                                                UDEMY EDA

Cargamos dataset udemy_courses.csv


In [273]:
data_udemy = pd.read_csv('./udemy_courses.csv')

In [274]:
data_udemy.head()

course_id                                       course_title  \
0    1070968                 Ultimate Investment Banking Course   
1    1113822  Complete GST Course & Certification - Grow You...   
2    1006314  Financial Modeling for Business Analysts and C...   
3    1210588  Beginner to Pro - Financial Analysis in Excel ...   
4    1011058       How To Maximize Your Profits Trading Options   

                                                 url  is_paid  price  \
0  https://www.udemy.com/ultimate-investment-bank...     True    200   
1      https://www.udemy.com/goods-and-services-tax/     True     75   
2  https://www.udemy.com/financial-modeling-for-b...     True     45   
3  https://www.udemy.com/complete-excel-finance-c...     True     95   
4  https://www.udemy.com/how-to-maximize-your-pro...     True    200   

   num_subscribers  num_reviews  num_lectures               level  \
0             2147           23            51          All Levels   
1             2792          923           274          All Levels   
2             2174           74            51  Intermediate Level   
3             2451           11            36          All Levels   
4             1276           45            26  Intermediate Level   

   content_duration   published_timestamp           subject  
0               1.5  2017-01-18T20:58:58Z  Business Finance  
1              39.0  2017-03-09T16:34:20Z  Business Finance  
2               2.5  2016-12-19T19:26:30Z  Business Finance  
3               3.0  2017-05-30T20:07:24Z  Business Finance  
4               2.0  2016-12-13T14:57:18Z  Business Finance

Análisis exploratorio para conocer la estructura de los datos y ver si hay errores o duplicados

In [22]:
data_udemy.info()
data_udemy.value_counts()
data_udemy.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3678 entries, 0 to 3677
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   course_id            3678 non-null   int64  
 1   course_title         3678 non-null   object 
 2   url                  3678 non-null   object 
 3   is_paid              3678 non-null   bool   
 4   price                3678 non-null   int64  
 5   num_subscribers      3678 non-null   int64  
 6   num_reviews          3678 non-null   int64  
 7   num_lectures         3678 non-null   int64  
 8   level                3678 non-null   object 
 9   content_duration     3678 non-null   float64
 10  published_timestamp  3678 non-null   object 
 11  subject              3678 non-null   object 
dtypes: bool(1), float64(1), int64(5), object(5)
memory usage: 319.8+ KB


course_id              0
course_title           0
url                    0
is_paid                0
price                  0
num_subscribers        0
num_reviews            0
num_lectures           0
level                  0
content_duration       0
published_timestamp    0
subject                0
dtype: int64

Cantidad de cursos pagos y cursos gratis.

In [23]:
data_udemy['is_paid'].value_counts() 

True     3368
False     310
Name: is_paid, dtype: int64

Cuento cuantos valores duplicados hay.

In [24]:
data_udemy.duplicated().value_counts() 

False    3672
True        6
dtype: int64

In [25]:
data_udemy_corregido = data_udemy.drop_duplicates(inplace=True)

Me quedo con las columnas que tengan relevancia para el análisis y las guardo en un duplicado para asi conservar el original.


In [26]:
data_udemy_corregido = data_udemy[['course_title','price','num_subscribers','level' ,'subject']]

In [27]:
data_udemy_corregido.head()

course_title  price  num_subscribers  \
0                 Ultimate Investment Banking Course    200             2147   
1  Complete GST Course & Certification - Grow You...     75             2792   
2  Financial Modeling for Business Analysts and C...     45             2174   
3  Beginner to Pro - Financial Analysis in Excel ...     95             2451   
4       How To Maximize Your Profits Trading Options    200             1276   

                level           subject  
0          All Levels  Business Finance  
1          All Levels  Business Finance  
2  Intermediate Level  Business Finance  
3          All Levels  Business Finance  
4  Intermediate Level  Business Finance

Reviso valores unicos de ciertas columnas

In [28]:
data_udemy_corregido['idioma'] = data_udemy_corregido['course_title'].apply(detectlang)

/var/folders/4n/z_dkq44j2t9bgj_4nq5l3m4h0000gn/T/ipykernel_2997/3012980803.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_udemy_corregido['idioma'] = data_udemy_corregido['course_title'].apply(detectlang)


In [29]:
data_udemy_corregido['price'].unique()

array([200,  75,  45,  95, 150,  65, 195,  30,  20,  50, 175, 140, 115,
       190, 125,  60, 145, 105, 155, 185, 180, 120,  25, 160,  40,   0,
       100,  90,  35,  80,  70,  55, 165, 130,  85, 170, 110, 135])

In [30]:
data_udemy_corregido['subject'].unique()

array(['Business Finance', 'Graphic Design', 'Musical Instruments',
       'Web Development'], dtype=object)

In [31]:
data_udemy_corregido['level'].unique()

array(['All Levels', 'Intermediate Level', 'Beginner Level',
       'Expert Level'], dtype=object)

Conteo de diferentes valores de la columna 'idioma'

In [138]:
data_udemy['idioma'].value_counts()

Ingles       3280
Español       140
Aleman         90
Portuges       66
Frances        44
Japones        29
Arabe           9
Chino           4
Turco           4
Ruso            3
Tailandes       2
Coreano         1
Name: idioma, dtype: int64

In [44]:
data_udemy = pd.read_csv('data_udemy_revisado.csv')


In [161]:
data_udemy['subject'].value_counts()

Web Development        1199
Business Finance       1191
Musical Instruments     680
Graphic Design          602
Name: subject, dtype: int64

In [139]:
data_udemy.to_csv('data_udemy_revisado.csv')

REALIZO UNA UNION DE LAS TABLAS PARA ESTUDIAR EL VALOR DE DIFICULTAD

Primero ejecuto algunas tranformaciones para normalizar la informacion de las columnas.

In [ ]:
data_coursera_final['Plataforma'] = 'Coursera'

In [281]:
data_udemy['Plataforma'] = 'Udemy'

In [284]:
data_edx_corregido['Plataforma'] = 'EDX'

In [294]:
data_coursera_final.rename({'name':'course_title'},axis=1,inplace=True)

/var/folders/4n/z_dkq44j2t9bgj_4nq5l3m4h0000gn/T/ipykernel_7871/1936717604.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_coursera_final.rename({'name':'course_title'},axis=1,inplace=True)


In [300]:
data_udemy.rename({'language':'idioma'},axis=1,inplace=True)

In [298]:
data_edx_corregido.rename({'language':'idioma'},axis=1,inplace=True)

Realizo copias con las columnas necesarias

In [302]:
data_coursera_final

course_title  \
0    Programming for Everybody (Getting Started wit...   
1                                     Machine Learning   
2                    Neural Networks and Deep Learning   
3                       Technical Support Fundamentals   
4    Learning How to Learn: Powerful mental tools t...   
..                                                 ...   
621            Digital Marketing Analytics in Practice   
622                    Dairy Production and Management   
623  Security & Safety Challenges in a Globalized W...   
624             The fundamentals of hotel distribution   
625                     A Crash Course in Data Science   

                            course_id  n_reviews  idioma Rating  \
0                              python    15226.0  Ingles    4.8   
1                    machine-learning    12677.0  Ingles    4.9   
2       neural-networks-deep-learning    12291.0  Ingles    4.9   
3      technical-support-fundamentals    12054.0  Ingles    4.8   
4               learning-how-to-learn    11871.0  Ingles    4.8   
..                                ...        ...     ...    ...   
621                 digital-analytics        0.0  Ingles      0   
622                  dairy-production        0.0  Ingles    4.8   
623  security-safety-globalized-world        0.0  Ingles    4.7   
624                hotel-distribution        0.0  Ingles    4.7   
625               data-science-course        0.0  Ingles    4.5   

             dificultad Plataforma  
0             All Level   Coursera  
1             All Level   Coursera  
2    Intermediate Level   Coursera  
3        Beginner Level   Coursera  
4             All Level   Coursera  
..                  ...        ...  
621           All Level   Coursera  
622           All Level   Coursera  
623      Beginner Level   Coursera  
624           All Level   Coursera  
625           All Level   Coursera  

[626 rows x 7 columns]

In [304]:
data_kpi_coursera = data_coursera_final[['course_title','dificultad','idioma','Plataforma']]

In [305]:
data_kpi_udemy = data_udemy[['course_title','dificultad','idioma','Plataforma']]

In [306]:
data_kpi_edx = data_edx_corregido[['course_title','dificultad','idioma','Plataforma']]

Ahora si union de los 3 datasets y las metricas que comparten para realizar el estudio de performance

In [312]:
data_level_moooc = data_kpi_coursera

In [320]:
data_level_moooc = pd.concat([data_kpi_coursera,data_kpi_edx,data_kpi_udemy])

In [321]:
data_level_moooc

course_title          dificultad  \
0     Programming for Everybody (Getting Started wit...           All Level   
1                                      Machine Learning           All Level   
2                     Neural Networks and Deep Learning  Intermediate Level   
3                        Technical Support Fundamentals      Beginner Level   
4     Learning How to Learn: Powerful mental tools t...           All Level   
...                                                 ...                 ...   
3667  Learn jQuery from Scratch - Master of JavaScri...          All Levels   
3668  How To Design A WordPress Website With No Codi...      Beginner Level   
3669                      Learn and Build using Polymer          All Levels   
3670  CSS Animations: Create Amazing Effects on Your...          All Levels   
3671  Using MODX CMS to Build Websites: A Beginner's...      Beginner Level   

      idioma Plataforma  
0     Ingles   Coursera  
1     Ingles   Coursera  
2     Ingles   Coursera  
3     Ingles   Coursera  
4     Ingles   Coursera  
...      ...        ...  
3667  Ingles      Udemy  
3668  Ingles      Udemy  
3669  Ingles      Udemy  
3670  Ingles      Udemy  
3671  Ingles      Udemy  

[5272 rows x 4 columns]

Exporto el dataset para usarlo

In [323]:
data_level_moooc.to_csv('data_mooc_kpi.csv')